### DISCUSSION TITLE
Instance UO inside Material

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#17418] [Q&A General] [answered]
Hello:

Beginner developer help needed here.  I want to instance a user object during the initialization of a material object.  To be specific, I have an ADMaterial "SPWallDragMaterial" that will produce a material property for wall drag.  I want the user to be able to select from a list of wall drag models.  So, I have a UO base class "WallDragModel", that has subclasses for the various models.

In SPWallDragMaterial.h, I have the declaration:

const WallDragModel & _wd;    // user object for wall drag model

In SPWallDragMaterial::validParams, there is a MooseEnum for the various models, along with the necessary input params that will need to be passed to the model UOs.  In the constructor for SPWallDragMaterial, I have:

  const std::string uo_name = this->name() + "_wall_drag";
  std::string uo_type = "WallDragModel";
  InputParameters params = WallDragModel::validParams();

  switch (_wall_drag_model)
  {
    case ASP::CONSTANT:
      uo_type = "ConstantWallDrag";
      params.set<Real>("value") = getParam<Real>("value");
      break;
     ..... other cases.....
  }

  _fe_problem.addUserObject(uo_type, uo_name, params);
  _wd = _fe_problem.getUserObject<WallDragModel>(uo_name);

I included "FEProblemBase.h".  I get the following compile error for the last line of code above (along with the error that I did not initialize _wd):

Users/joe/projects/ASP/src/materials/SPWallDragMaterial.C:82:7: error: no viable overloaded '='
  _wd = _fe_problem.getUserObject<WallDragModel>(uo_name);
  ~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/joe/projects/ASP/build/header_symlinks/WallDragModel.h:5:7: note: candidate function (the implicit copy assignment operator) not viable: 'this' argument has type 'const WallDragModel', but method is not marked const
class WallDragModel : public GeneralUserObject

What is the proper way to instance a UO inside of a Material class?  

Thanks for the help,
  Joe Kelly

**Comment by ngrilli** [comment]
Dear Joe Kelly,

The way I use user objects inside materials is shown in this example: 
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/ComputeElasticityTensorCPGrain.C

In the class parameters you add:
  params.addParam<UserObjectName>("read_prop_user_object",
                                  "The GrainPropertyReadFile "
                                  "GeneralUserObject to read element "
                                  "specific property values from file");

Then in the constructor:
    _read_prop_user_object(isParamValid("read_prop_user_object")
                               ? &getUserObject<GrainPropertyReadFile>("read_prop_user_object")
                               : nullptr),

Then methods of the user object can be called as:
_read_prop_user_object->getData(_current_elem, 0);

In the header file, there is:
const GrainPropertyReadFile * const _read_prop_user_object;

In this case you don't need the addUserObject function.

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by ngrilli** [reply]
@joe61vette

**Reply by joe61vette** [reply]
Dear Nicolo:

Thanks so much for your reply.  If I understand this correctly, the "read_prop_user_object" would have to already exist.  Namely be part of the input file.  Thus it would already be instanced and initialized.  The coding you provided would then "use" that UO.  Is that correct?

What I am hoping to do is actually instance the UO inside the material.  So, no UO in the input model.  The input params would actually be read by the material and then passed to the UO when it is instanced.  I would like to follow this approach as a user convenience feature.

Thanks for the help,
  Joe

**Comment by rwcarlsen** [comment, marked as answer]
Would it be better to maybe just make all these things materials?  You could create a material for each drag model - e.g. FooDragModel, BarDragModel, etc.  Then you could have this material create properties with its own name plus a suffix.  The wall drag material itself would then take the name of the drag model material as a parameter - and then add the pre-coded suffixes to retrieve the appropriate values supplied by the drag model.  Something like this:

```
[Materials]
  [drag_model]
    type = FooDragModel
    # this material creates properties named "drag_model_prop1", "drag_model_prop2" - etc.
  []
  [wall_drag]
    type = WallDrag
    model = drag_model # uses this suffix to calc+retrieve properties "drag_model_prop1", "drag_model_prop2", etc.
  []
[]
```

**Reply by joe61vette** [reply]
Thanks Robert.  I think that would work.  I will need to look at it as I would like to instance the drag_model specific materials from the wall_drag material rather than have the user need to put them in the input model.  Wasn't sure if I needed an action for this or not.  In some of the drag_models, there is a significant amount of setup and data to store.  That is why I want them in their own objects.

Joe

**Reply by rwcarlsen** [reply]
If you really want to streamline input file syntax, then an action that sets things up would be the way to go.  MOOSE really isn't designed for MooseObject/system subclasses (nearly all classes/objects in moose) to be created ouside of input files and/or actions.  Anything else you try will just be a bit wonky and won't be a "first-class" work-flow.  It's also generally good to try to stick with materials as often as reasonably doable over other objects (userobjects, etc.) - they offer some power/flexibility that can be quite convenient as your code evolves.
